prow/config/config.go

```diff
@@ -962,52 +948,59 @@ func (c *JobConfig) decorationRequested() bool {
 	return false
 }
 
-func validateLabels(name string, labels map[string]string) error {
-	for label := range labels {
+func validateLabels(labels map[string]string) error {
+	for label, value := range labels {
 		for _, prowLabel := range decorate.Labels() {
 			if label == prowLabel {
-				return fmt.Errorf("job %s attempted to set Prow-controlled label %s to %s", name, label, labels[label])
+				return fmt.Errorf("label %s is reserved for decoration", label)
 			}
 		}
 		if errs := validation.IsQualifiedName(label); len(errs) != 0 {
-			return fmt.Errorf("job %s sets an invalid label key %s: %v", name, label, errs)
+			return fmt.Errorf("invalid label %s: %v", label, errs)
 		}
 		if errs := validation.IsValidLabelValue(labels[label]); len(errs) != 0 {
-			return fmt.Errorf("job %s sets an invalid label value %s for key %s: %v", name, labels[label], label, errs)
+			return fmt.Errorf("label %s has invalid value %s: %v", label, value, errs)
 		}
 	}
 	return nil
 }
 
-func validateAgent(name, agent string, spec *v1.PodSpec, config *kube.DecorationConfig) error {
-	// Ensure that k8s jobs have a pod spec.
-	if agent == string(kube.KubernetesAgent) && spec == nil {
-		return fmt.Errorf("job %s has no spec", name)
-	}
-	// Only k8s jobs can be decorated
-	if agent != string(kube.KubernetesAgent) && config != nil {
-		return fmt.Errorf("job %s configured PodSpec decoration but is not a Kubernetes job", name)
-	}
-	// Ensure agent is a known value.
-	if agent != string(kube.KubernetesAgent) && agent != string(kube.JenkinsAgent) {
-		return fmt.Errorf("job %s has invalid agent (%s), it needs to be one of the following: %s %s",
-			name, agent, kube.KubernetesAgent, kube.JenkinsAgent)
+func validateAgent(v JobBase) error {
+	k := string(prowjobv1.KubernetesAgent)
+	b := string(prowjobv1.KnativeBuildAgent)
+	j := string(prowjobv1.JenkinsAgent)
```

Comment:
to future test-infra: if we get n > 3 I think we should probably define the list of valid agents somewhere common.